File: src/ShoppingCart.Application/Services/ProductService.cs
using Mapster;
using Microsoft.EntityFrameworkCore;
using ShoppingCart.Application.DTOs.Products;
using ShoppingCart.Application.Interfaces.Services;
using ShoppingCart.Domain.Entities;
using ShoppingCart.Persistence.Data;
using ShoppingCart.Shared.Common;
using ShoppingCart.Shared.Exceptions;

namespace ShoppingCart.Application.Services;

public class ProductService : IProductService
{
    private readonly AppDbContext _context;

    public ProductService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<PagedResult<ProductListDto>> GetProductsAsync(
        int pageIndex, int pageSize, string? search, int? categoryId, int? brandId, string? sortBy, bool? isFeatured, bool? isBestSeller = null, bool? isNewArrival = null)
    {
        var query = _context.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .Where(p => p.IsActive)
            .AsQueryable();

        if (!string.IsNullOrEmpty(search))
        {
            query = query.Where(p => p.Name.Contains(search) || 
                                     (p.SKU != null && p.SKU.Contains(search)));
        }

        if (categoryId.HasValue)
        {
            query = query.Where(p => p.CategoryId == categoryId.Value);
        }

        if (brandId.HasValue)
        {
            query = query.Where(p => p.BrandId == brandId.Value);
        }

        if (isFeatured.HasValue && isFeatured.Value)
        {
            query = query.Where(p => p.IsFeatured);
        }

        if (isBestSeller.HasValue && isBestSeller.Value)
        {
            query = query.Where(p => p.IsBestSeller);
        }

        if (isNewArrival.HasValue && isNewArrival.Value)
        {
            query = query.Where(p => p.IsNewArrival);
        }

        query = sortBy?.ToLower() switch
        {
            "price_asc" => query.OrderBy(p => p.Price),
            "price_desc" => query.OrderByDescending(p => p.Price),
            "name_asc" => query.OrderBy(p => p.Name),
            "name_desc" => query.OrderByDescending(p => p.Name),
            "newest" => query.OrderByDescending(p => p.CreatedAt),
            _ => query.OrderByDescending(p => p.IsFeatured).ThenByDescending(p => p.CreatedAt)
        };

        var totalCount = await query.CountAsync();
        var items = await query
            .Skip(pageIndex * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return new PagedResult<ProductListDto>(
            items.Adapt<List<ProductListDto>>(),
            totalCount,
            pageIndex,
            pageSize);
    }

    public async Task<ProductDto?> GetProductByIdAsync(int id)
    {
        var product = await _context.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images.OrderBy(i => i.DisplayOrder))
            .Include(p => p.Variants.Where(v => v.IsActive))
            .Include(p => p.StockItems)
            .FirstOrDefaultAsync(p => p.Id == id);

        return product?.Adapt<ProductDto>();
    }

    public async Task<ProductDto?> GetProductBySlugAsync(string slug)
    {
        var product = await _context.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images.OrderBy(i => i.DisplayOrder))
            .Include(p => p.Variants.Where(v => v.IsActive))
            .Include(p => p.StockItems)
            .FirstOrDefaultAsync(p => p.Slug == slug && p.IsActive);

        return product?.Adapt<ProductDto>();
    }

    public async Task<ProductDto> CreateProductAsync(CreateProductRequest request)
    {
        var product = request.Adapt<Product>();
        product.Slug = GenerateUniqueSlug(request.Name);
        product.CreatedAt = DateTime.UtcNow;
        product.CreatedBy = "system";

        _context.Products.Add(product);
        await _context.SaveChangesAsync();

        if (request.Images?.Any() == true)
        {
            foreach (var imageRequest in request.Images)
            {
                var image = imageRequest.Adapt<ProductImage>();
                image.ProductId = product.Id;
                image.CreatedAt = DateTime.UtcNow;
                image.CreatedBy = "system";
                _context.ProductImages.Add(image);
            }
            await _context.SaveChangesAsync();
        }

        return (await GetProductByIdAsync(product.Id))!;
    }

    public async Task<ProductDto> UpdateProductAsync(UpdateProductRequest request)
    {
        var product = await _context.Products
            .Include(p => p.Images)
            .FirstOrDefaultAsync(p => p.Id == request.Id);

        if (product == null)
        {
            throw new NotFoundException("Product not found");
        }

        product.Name = request.Name;
        product.SKU = request.SKU;
        product.Barcode = request.Barcode;
        product.ShortDescription = request.ShortDescription;
        product.Description = request.Description;
        product.Price = request.Price;
        product.OldPrice = request.OldPrice;
        product.CostPrice = request.CostPrice;
        product.CategoryId = request.CategoryId;
        product.BrandId = request.BrandId;
        product.SupplierId = request.SupplierId;
        product.IsFeatured = request.IsFeatured;
        product.IsBestSeller = request.IsBestSeller;
        product.IsNewArrival = request.IsNewArrival;
        product.MinimumStockLevel = request.MinimumStockLevel;
        product.ReorderLevel = request.ReorderLevel;
        product.Weight = request.Weight;
        product.Dimensions = request.Dimensions;
        product.IsActive = request.IsActive;
        product.UpdatedAt = DateTime.UtcNow;
        product.UpdatedBy = "system";

        await _context.SaveChangesAsync();
        return (await GetProductByIdAsync(product.Id))!;
    }

    public async Task<bool> DeleteProductAsync(int id)
    {
        var product = await _context.Products.FindAsync(id);
        if (product == null)
        {
            return false;
        }

        product.IsDeleted = true;
        product.UpdatedAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<bool> ToggleFeaturedAsync(int id)
    {
        var product = await _context.Products.FindAsync(id);
        if (product == null)
        {
            return false;
        }

        product.IsFeatured = !product.IsFeatured;
        product.UpdatedAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<List<ProductListDto>> GetFeaturedProductsAsync(int count = 10)
    {
        var products = await _context.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .Where(p => p.IsActive && p.IsFeatured)
            .OrderByDescending(p => p.CreatedAt)
            .Take(count)
            .ToListAsync();

        return products.Adapt<List<ProductListDto>>();
    }

    public async Task<List<ProductListDto>> GetBestSellersAsync(int count = 10)
    {
        var productIds = await _context.OrderItems
            .GroupBy(oi => oi.ProductId)
            .OrderByDescending(g => g.Sum(oi => oi.Quantity))
            .Take(count)
            .Select(g => g.Key)
            .ToListAsync();

        var products = await _context.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .Where(p => productIds.Contains(p.Id) && p.IsActive)
            .ToListAsync();

        return products.Adapt<List<ProductListDto>>();
    }

    public async Task<List<ProductListDto>> GetNewArrivalsAsync(int count = 10)
    {
        var products = await _context.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .Where(p => p.IsActive && p.IsNewArrival)
            .OrderByDescending(p => p.CreatedAt)
            .Take(count)
            .ToListAsync();

        return products.Adapt<List<ProductListDto>>();
    }

    public async Task<List<ProductListDto>> GetRelatedProductsAsync(int productId, int count = 8)
    {
        var product = await _context.Products.FindAsync(productId);
        if (product == null)
        {
            return new List<ProductListDto>();
        }

        var products = await _context.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .Where(p => p.Id != productId && 
                        p.CategoryId == product.CategoryId && 
                        p.IsActive)
            .OrderByDescending(p => p.IsFeatured)
            .ThenByDescending(p => p.CreatedAt)
            .Take(count)
            .ToListAsync();

        return products.Adapt<List<ProductListDto>>();
    }

    public async Task<List<ProductListDto>> SearchProductsAsync(string query, int count = 20)
    {
        var products = await _context.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .Where(p => p.IsActive && 
                        (p.Name.Contains(query) || 
                         (p.SKU != null && p.SKU.Contains(query))))
            .Take(count)
            .ToListAsync();

        return products.Adapt<List<ProductListDto>>();
    }

    public async Task IncrementViewCountAsync(int productId)
    {
        var product = await _context.Products.FindAsync(productId);
        if (product != null)
        {
            product.ViewCount++;
            await _context.SaveChangesAsync();
        }
    }

    private string GenerateUniqueSlug(string name)
    {
        var slug = name.ToLowerInvariant()
            .Replace(" ", "-")
            .Replace("'", "")
            .Replace("\"", "");

        var existingCount = _context.Products
            .Count(p => p.Slug.StartsWith(slug));

        return existingCount > 0 ? $"{slug}-{existingCount + 1}" : slug;
    }
}

File: frontend/shoppingcart-app/src/services/productService.ts
import api from './api'

export interface Product {
  id: number
  name: string
  slug: string
  sku?: string
  price: number
  oldPrice?: number
  shortDescription?: string
  description?: string
  categoryName: string
  brandName?: string
  isFeatured: boolean
  isBestSeller: boolean
  isNewArrival: boolean
  stockQuantity: number
  stockStatus: string
  ratingAverage: number
  reviewCount: number
  viewCount: number
  images: { id: number; imageUrl: string; isMain: boolean }[]
  mainImageUrl?: string
  variants?: ProductVariant[]
}

export interface ProductVariant {
  id: number
  name: string
  sku?: string
  price: number
  stockQuantity: number
  imageUrl?: string
  isActive: boolean
  attributeValues: { attributeId: number; attributeName: string; value: string }[]
}

export interface CreateVariantRequest {
  productId: number
  name: string
  sku?: string
  barcode?: string
  price: number
  costPrice: number
  stockQuantity: number
  imageUrl?: string
  isActive: boolean
  attributes: { attributeId: number; value: string }[]
}

export interface Category {
  id: number
  name: string
  slug: string
  iconUrl?: string
  imageUrl?: string
  isFeatured: boolean
  productCount: number
}

export interface PagedResult<T> {
  items: T[]
  totalCount: number
  pageIndex: number
  pageSize: number
  totalPages: number
  hasPrevious: boolean
  hasNext: boolean
}

export interface CreateProductRequest {
  name: string
  sku?: string
  barcode?: string
  shortDescription?: string
  description?: string
  price: number
  oldPrice?: number
  costPrice: number
  categoryId: number
  brandId?: number
  supplierId?: number
  isFeatured: boolean
  isBestSeller: boolean
  isNewArrival: boolean
  minimumStockLevel: number
  reorderLevel: number
  weight: number
  dimensions?: string
  images?: { imageUrl: string; altText?: string; displayOrder: number; isMain: boolean }[]
}

export const productService = {
  getProducts: async (params?: {
    pageIndex?: number
    pageSize?: number
    search?: string
    categoryId?: number
    brandId?: number
    sortBy?: string
    isFeatured?: boolean
    isBestSeller?: boolean
    isNewArrival?: boolean
  }): Promise<PagedResult<Product>> => {
    const { data } = await api.get('/products', { params })
    return data.data
  },

  createProduct: async (product: CreateProductRequest): Promise<Product> => {
    const { data } = await api.post('/products', product)
    return data.data
  },

  getProduct: async (idOrSlug: string, isSlug = false): Promise<Product> => {
    const url = isSlug ? `/products/slug/${idOrSlug}` : `/products/${idOrSlug}`
    const { data } = await api.get(url)
    return data.data
  },

  getFeaturedProducts: async (count = 10): Promise<Product[]> => {
    const { data } = await api.get('/products/featured', { params: { count } })
    return data.data
  },

  getBestSellers: async (count = 10): Promise<Product[]> => {
    const { data } = await api.get('/products/best-sellers', { params: { count } })
    return data.data
  },

  getNewArrivals: async (count = 10): Promise<Product[]> => {
    const { data } = await api.get('/products/new-arrivals', { params: { count } })
    return data.data
  },

  getRelatedProducts: async (productId: number, count = 8): Promise<Product[]> => {
    const { data } = await api.get(`/products/${productId}/related`, { params: { count } })
    return data.data
  },

  searchProducts: async (query: string, count = 20): Promise<Product[]> => {
    const { data } = await api.get('/products/search', { params: { query, count } })
    return data.data
  },
}

export const variantService = {
  getVariantsByProduct: async (productId: number): Promise<ProductVariant[]> => {
    const { data } = await api.get(`/variants/product/${productId}`)
    return data.data
  },

  getVariant: async (id: number): Promise<ProductVariant> => {
    const { data } = await api.get(`/variants/${id}`)
    return data.data
  },

  createVariant: async (variant: CreateVariantRequest): Promise<ProductVariant> => {
    const { data } = await api.post('/variants', variant)
    return data.data
  },

  updateVariant: async (id: number, variant: Omit<CreateVariantRequest, 'productId'>): Promise<ProductVariant> => {
    const { data } = await api.put(`/variants/${id}`, variant)
    return data.data
  },

  deleteVariant: async (id: number): Promise<void> => {
    await api.delete(`/variants/${id}`)
  },
}

export const categoryService = {
  getCategories: async (): Promise<Category[]> => {
    const { data } = await api.get('/categories/all')
    return data.data
  },

  getFeaturedCategories: async (): Promise<Category[]> => {
    const { data } = await api.get('/categories/featured')
    return data.data
  },

  getCategory: async (idOrSlug: string, isSlug = false): Promise<Category> => {
    const url = isSlug ? `/categories/slug/${idOrSlug}` : `/categories/${idOrSlug}`
    const { data } = await api.get(url)
    return data.data
  },
}
